### ISSUE TITLE
Sync reports being in Idle state when it is still in the middle of syncing

### ISSUE THREAD

**Issue by jcamp1** [couchbase/couchbase-lite-java-core#456]
Greetings,

Sync will report being in an idle state with 0/0 documents being synced in the middle of syncing. For example:

02-24 10:48:46.312    2856-2890/com.app.app.alpha D/App﹕ Progress: 149 / 586
02-24 10:48:46.312    2856-2890/com.app.app.alpha D/App﹕ ----------->Sync Session: active: true status: Active 149/586 Error (null)
02-24 10:48:46.382    2856-2890/com.app.app.alpha D/App﹕ Progress: 0 / 0
02-24 10:48:46.382    2856-2890/com.app.app.alpha D/App﹕ ----------->Sync Session: active: false status: Idle 0/0 Error (null)

This makes it impossible to know when an initial sync is complete and doesn't mirror the iOS behavior.

Cheers,
Justin

**Comment by jcamp1** [comment]
Upon further investigation, this appears to be the push replicator reporting that it has nothing to do. Is this the correct behavior? I believe in iOS when the push replicator has nothing to do, it doesn't report anything.

**Comment by hideki** [comment]
With continuous mode, both push and pull replicators notify ACTIVE if there is documents which are needed to be replicated, and notify IDLE when replication is completed.
Also CBL iOS and Android replicator behavior should be compatible.

**Comment by jcamp1** [comment]
Hi @hideki,

Yes, this is understood. That's not the scenario I'm in however. In a fully empty database with a fresh sync starting, we're opening both a pull and a push replicator. During the time that the pull replicator is syncing all of the documents down, the push replicator occasionally emits it's "I have nothing to do" callback (which is obvious as we're syncing new documents down to a fresh database). This is not the behavior we're seeing in iOS.

It's not a huge problem, as I can simply stop listening to the push replicator during initial sync, but it doesn't seem to be correct behavior.

Thanks!
Justin

**Comment by hideki** [comment]
@jcamp1 Thank you for your additional information. I will look into the issue. But I have some open tickets I need to take care before.

**Comment by tleyden** [comment]
Investigate and write a unit test

**Comment by hideki** [comment]
Hi @jcamp1,

I could not reproduce this issue with my test app.

The test app creates continuous push replicator when it starts, and it logs if `Replication.ChangeEvent` is called. Link to simple test app. https://github.com/hideki/couchbase-lite/tree/master/test/Issue456

From below log message, Push replicator became ACTIVE state just after it started, then immediately became IDLE state because nothing necessary to push. After than, `Replication.ChangeEvent` is never called. 

Log message
```
03-24 21:09:39.301    1598-1613/self.hideki.issue456 E/Issue456﹕ PUSH replication event. Source: com.couchbase.lite.replicator.Replication@52864654 Transition: INITIAL -> RUNNING Total changes: 0 Completed changes: 0
03-24 21:09:39.493    1598-1613/self.hideki.issue456 E/Issue456﹕ PUSH replication event. Source: com.couchbase.lite.replicator.Replication@52864654 Transition: RUNNING -> IDLE Total changes: 0 Completed changes: 0
```

Note: If Push replicator becomes IDLE during Pull replicator is replicating, your code needs to check which replicator becomes IDLE state.

I will close this ticket. If I mis-understand your bug report, please reopen this ticket?

Thanks!

**Comment by jcamp1** [comment]
Hi @hideki,

This isn't the behavior I was seeing, but I believe it has to do with having two replicators running that are both set to continuous on an initial sync. This will probably be worked out while looking into #498.

Cheers!

**Comment by hideki** [comment]
I reproduced the issue of Replicator state becomes IDLE middle of replication with continuous mode. Related with #498 .

**Comment by hideki** [comment]
Trigger by following line of code
https://github.com/couchbase/couchbase-lite-java-core/blob/c98ad8ec3b97fcf22257755cd6424e1cc1fa4300/src/main/java/com/couchbase/lite/replicator/PullerInternal.java#L995

```
04-16 17:55:25.265   9811-10654/self.hideki.issue4982a D/Sync﹕ done calling future.get() on java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@528933bc
04-16 17:55:25.269   9811-10654/self.hideki.issue4982a W/Sync﹕ [fireTrigger()] => WAITING_FOR_CHANGES
04-16 17:55:25.269   9811-10654/self.hideki.issue4982a W/System.err﹕ java.lang.Exception: WAITING_FOR_CHANGES
04-16 17:55:25.269   9811-10654/self.hideki.issue4982a W/System.err﹕ at com.couchbase.lite.replicator.ReplicationInternal.fireTrigger(ReplicationInternal.java:189)
04-16 17:55:25.269   9811-10654/self.hideki.issue4982a W/System.err﹕ at com.couchbase.lite.replicator.PullerInternal$13.run(PullerInternal.java:995)
04-16 17:55:25.269   9811-10654/self.hideki.issue4982a W/System.err﹕ at java.lang.Thread.run(Thread.java:841)
```

**Comment by hideki** [comment]
Hi @jcamp1,

We also fixed this issue. Could you please test with your environment?

Fix is in release/1.1.0 branch, we will merge this into master after 1.1.0 release.
https://github.com/couchbase/couchbase-lite-java-core/tree/release/1.1.0
